### PR TITLE
feat(cli): added jest tests to typescript templates

### LINKED ISSUE #3945: cdk init lib should come with unit tests
### General Information

* [X] :wave: I may be able to implement this feature request

* [ ] :warning: This feature might incur a breaking change

### Description
when running `cdk init --language typescript lib` the generated code should include the devDependencies on nodeunit and cdk-assert as well as sample unit tests for the SQS Queue and SNS Topic it creates by default.

### Proposed Solution

Should be a pretty straight-forward change to the TS Template.
### Environment
  - **Language:** TypeScript

### PR DESCRIPTION
**Pull request by rhboyd**
Added tests based on the Jest test framework for testing the lib Constructs

closes #3945
----
Two simple tests that ensure the SQS Queue and the SNS Topic are created.

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by jogold** [comment]
Suggestion: show how to use snapshot testing of the whole stack with jest

**Comment by rhboyd** [comment]
@eladb that’s a good question. I just didn’t think to add it to the other two templates. I’ll update the PR.

**Comment by rhboyd** [comment]
`app` doesn't include any Constructs by default. Should I add an empty test that just builds the empty Stack or should I add a Construct to the app template?

`sample-app` should be pretty straight-forward

**Comment by rhboyd** [comment]
@jogold I'm pretty new to testing in TS/Jest, do you have an example or guide for snapshot testing in general with these tools?

**Comment by eladb** [comment]
@rhboyd i think a test that asserts that the stack produces an empty template is good

**Comment by rhboyd** [comment]
Added the tests to `app` and `sample-app` as well. I tested this by running the local version of the CDK CLI to generate projects from each template and running `npm install && npm run build && npm run test` to ensure that the tests ran successfully.

**Comment by rhboyd** [comment]
@jogold Is this the type of snapshot testing you were looking for?

https://github.com/aws/aws-cdk/issues/300

File: packages/aws-cdk/lib/init-templates/app/typescript/package.template.json [hunk 1]
```diff
@@ -7,9 +7,14 @@
   "scripts": {
     "build": "tsc",
     "watch": "tsc -w",
+    "test": "jest",
     "cdk": "cdk"
   },
   "devDependencies": {
+    "@aws-cdk/assert": "^1.6.1",
```
Comment (jogold) [inline packages/aws-cdk/lib/init-templates/app/typescript/package.template.json]:
```suggestion
    "@aws-cdk/assert": "^%cdk-version%",
```

**Comment by jogold** [comment]
> @jogold Is this the type of snapshot testing you were looking for?
> 
> #300

Yes, using `SynthUtils` from `@aws-cdk/assert` you can do:
```ts
import { SynthUtils } from '@aws-cdk/assert';
import cdk = require('@aws-cdk/core');
import initCdkStack = require('../lib/init-cdk-stack');

test('Stack Snapshot', () => {
  const app = new cdk.App();

  const stack = new hello.HelloStack(app, 'Test');
  
  expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot('HelloStack');
});
```

Running `jest` the first time will create a snapshot of your stack. This snapshot is committed with the code and can be reviewed. When your code changes, you run `jest -u` to update your snapshot. The changes in the snapshot can be code reviewed. Also, when updating the CDK deps you can be confident that your CF template remains stable (prevents infrastructure breaking changes) and if not, clearly view what will change in your infrastructure.

I think that this is the best way to test a stack. For me, testing with `haveResource` is like re-testing what's already tested in the CDK.

**Comment by jogold** [comment]
Without a `jest.config.js` your are testing both `.js` and `.ts` files here, see default for `testMatch` https://jestjs.io/docs/en/configuration#testmatch-array-string

**Comment by rhboyd** [comment]
> Without a `jest.config.js` (or equivalent) you are testing both `.js` and `.ts` files here, see default for `testMatch` https://jestjs.io/docs/en/configuration#testmatch-array-string

ohhh, the jest.config.js file was excluded by the gitignore. fixing it now

File: packages/aws-cdk/lib/init-templates/app/typescript/package.template.json (hunk 1 above)
Comment (rhboyd) [reply packages/aws-cdk/lib/init-templates/app/typescript/package.template.json]: fixed

File: packages/aws-cdk/lib/init-templates/lib/typescript/jest.config.js [hunk 2]
```diff
@@ -0,0 +1,8 @@
+module.exports = {
```
Comment (jogold) [inline packages/aws-cdk/lib/init-templates/lib/typescript/jest.config.js]: You still have to specify `testMatch: [ '**/*.test.ts']` because after `npm run build` you will end up with `.js` files in your test folder.

I would also add `testEnvironment: 'node'` + may be coverage ?

Comment (rhboyd) [reply packages/aws-cdk/lib/init-templates/lib/typescript/jest.config.js]: Added the testMatch. I don't think we need to get into adding test coverage for the initial templates.

**Comment by rhboyd** [comment]
Will do.

**Comment by rhboyd** [comment]
Added the command to run `npm run test` during cli integration test and manually re-ran CLI Integration Tests. (submitted a separate PR for an unrelated broken Integ Test)
